sky/templates/websocket_proxy.py

```diff
@@ -75,45 +82,134 @@ async def main(url: str) -> None:
         print(f'Error in stdin_to_websocket: {e}', file=sys.stderr)
     finally:
         await websocket.close()
+        websocket_closed_event.set()
 
 
 async def websocket_to_stdout(websocket: ClientConnection,
-                              writer: asyncio.StreamWriter):
+                              writer: asyncio.StreamWriter,
+                              timestamps_supported: bool,
+                              last_ping_time_dict: Optional[dict],
+                              websocket_closed_event: asyncio.Event):
     try:
-        while True:
+        while not websocket_closed_event.is_set():
             message = await websocket.recv()
-            writer.write(message)
-            await writer.drain()
+            if timestamps_supported and len(message) > 0:
+                message_type = struct.unpack('!B', message[:1])[0]
+                if message_type == KubernetesSSHMessageType.REGULAR_DATA.value:
+                    # Regular data - strip type byte and write to stdout
+                    data = message[1:]
+                    writer.write(data)
+                    await writer.drain()
+                elif message_type == KubernetesSSHMessageType.PINGPONG.value:
+                    # PONG response - calculate latency and send measurement
+                    pong_time = time.time()
+                    if last_ping_time_dict and 'time' in last_ping_time_dict:
+                        ping_time = last_ping_time_dict['time']
+                        latency_seconds = pong_time - ping_time
+                        latency_ms = int(latency_seconds * 1000)
+
+                        # Send latency measurement (type 2)
+                        message_type_bytes = struct.pack(
+                            '!B',
+                            KubernetesSSHMessageType.LATENCY_MEASUREMENT.value)
+                        latency_bytes = struct.pack('!Q', latency_ms)
+                        data = message_type_bytes + latency_bytes
+                        try:
+                            await websocket.send(data)
+                        except Exception as e:  # pylint: disable=broad-except
+                            print(f'Failed to send latency measurement: {e}',
+                                  file=sys.stderr)
+                else:
+                    # Unknown message type, write as-is
+                    writer.write(message)
+                    await writer.drain()
+            else:
+                # No timestamps support, write directly
+                writer.write(message)
+                await writer.drain()
     except websockets.exceptions.ConnectionClosed:
         print('WebSocket connection closed', file=sys.stderr)
     except Exception as e:  # pylint: disable=broad-except
         print(f'Error in websocket_to_stdout: {e}', file=sys.stderr)
+        raise e
+    finally:
+        await websocket.close()
+        websocket_closed_event.set()
```

Comment:
By having this in both places, won't we try to close the websocket twice? Was this needed to fix some other issue?